torch/fx/subgraph_rewriter.py

```diff
@@ -0,0 +1,202 @@
+            if not match_found:
+                self.nodes_map.pop(n1)
+                return False
+
+        return True
+
+
+def replace_pattern(original_graph: Graph, pattern : Callable, replacement : Callable) -> None:
+    """
+    Function that takes one set of operators and their data dependencies
+    and replaces them with another
+    """
+
+    from copy import copy
+
+    # Get the graphs for `pattern` and `replacement`
+    pattern_graph = symbolic_trace(pattern).graph
+    replacement_graph = symbolic_trace(replacement).graph
+
+    # Find all possible pattern matches in original_graph. Note that pattern
+    # matches may overlap with each other.
+    matcher = SubgraphMatcher(pattern_graph)
+    matches: List[Match] = []
+
+    # Consider each node as an "anchor" (deepest matching graph node)
+    for anchor in original_graph.nodes:
+
+        if matcher.matches_subgraph_from_anchor(anchor):
+
+            def pattern_is_contained(nodes_map: Dict[Node, Node]) -> bool:
+                # `lookup` represents all the nodes in `original_graph`
+                # that are part of `pattern`
+                lookup: Dict[Node, Node] = {v : k for k, v
+                                            in nodes_map.items()}
+                for n in lookup.keys():
+                    if n.op == "placeholder" or lookup[n].op == "output":
+                        continue
+                    for user in n.users:
+                        # If this node has users that were not in
+                        # `lookup`, then it must leak out of the
+                        # pattern subgraph
+                        if user not in lookup:
+                            return False
+                return True
+
+            # It's not a match if the pattern leaks out into the rest
+            # of the graph
+            if pattern_is_contained(matcher.nodes_map):
+                # Shallow copy nodes_map
+                matches.append(Match(anchor=anchor,
+                                     nodes_map=copy(matcher.nodes_map)))
+
+    # The set of all nodes in `original_graph` that we've seen thus far
+    # as part of a pattern match
+    nodes_to_delete: Set[Node] = set()
+
+    # Get the Nodes we'll use to hook the replacement subgraph in to the
+    # original graph
+    replacement_inputs: List[Node] = [n for n in replacement_graph.nodes
+                                      if n.op == "placeholder"]
+    replacement_output: Node = next(n for n in replacement_graph.nodes
+                                    if n.op == "output")
+
+    # Return TRUE if one of the nodes in the current match has already
+    # been used as part of another match
+    def overlaps_with_prev_match(match: Match) -> bool:
+        for n in match.nodes_map.values():
+            if n in nodes_to_delete:
+                return True
+        return False
+
+    if not matches:
+        raise ValueError("No matches were found for the given pattern")
```

Comment:
This should just return early. You can imagine someone is making a pipeline of replacements that may or may not exist in the model we're passing through (e.g. conv-bn folding, then linear-activation folding, etc). It would be more expected for this to just return rather than throw if the pattern isn't there